# File: 06_samtools_array_job.R
# Auth: umar.niazi@kcl.as.uk
# DESC: create a parameter file and shell script to run array job on hpc
# Date: 4/9/2022


## set variables and source libraries
source('header.R')

## connect to mysql database 
library('RMySQL')

##### connect to mysql database to get samples
db = dbConnect(MySQL(), user='rstudio', password='12345', dbname='Projects', host='127.0.0.1')
dbListTables(db)

# load file data
dbGetQuery(db, paste('describe File;'))
dbGetQuery(db, paste('describe Sample;'))

g_did
dfSamples = dbGetQuery(db, paste('select File.*, Sample.id, Sample.idData from File, Sample where File.idSample = Sample.id having Sample.idData = 54;'))

f = dfSamples$group1 %in% 'Aligned by sequencing centre'
dfSamples = dfSamples[f,]
dbDisconnect(db)

csFiles = dfSamples$name

# set header variables 
cvShell = '#!/bin/bash -l'
cvJobName = '#SBATCH --job-name=samtools-array'
cvNodes = '#SBATCH --nodes=1'
cvProcessors = '#SBATCH --ntasks=1'
# format d-hh:mm:ss
cvRuntime = '#SBATCH --time=2-00:00:00'
cvPartition = '#SBATCH --partition=cpu'
# How much memory you need.
# --mem will define memory per node and
# --mem-per-cpu will define memory per CPU/core. Choose one of those.
cvMemoryReserve = '#SBATCH --mem-per-cpu=5000MB'
# Turn on mail notification. There are many possible self-explaining values:
# NONE, BEGIN, END, FAIL, ALL (including all aforementioned)
# For more values, check "man sbatch"
cvMail = '#SBATCH --mail-type=END,FAIL'
# set array job loop
length(csFiles)
cvArrayJob = '#SBATCH --array=1-73'

# set the directory names
cvInput = 'input/'
cvOutput = 'output/'
#cvSam = 'samtools'

# create a parameter file and shell script
dir.create('AutoScripts')
oFile.param = file('AutoScripts/samtools_param.txt', 'wt')
csFiles = gsub('.bam', '', csFiles)
# csFiles = gsub('trim_', '', csFiles)

temp = lapply(seq_along(csFiles), function(x){
  lf = csFiles[x]
  # write samtools command variables
  in.s1 = paste0(cvInput,lf, '.bam')
  # output file in bam format
  # s2 = paste0(cvInput, lf, '.bam')
  # remove MAPQ quality reads below 30
  s3 = paste0(cvOutput, lf, '_q30.bam')
  # sort the file by name i.e. the QNAME field
  s4 = paste0(cvOutput, lf, '_q30_sortn.bam')
  # fix mate
  s5 = paste0(cvOutput, lf, '_q30_sortn_fixm.bam')
  # sort by position
  s6 = paste0(cvOutput, lf, '_q30_fixm_sortc.bam')
  # remove or mark duplicates
  s7 = paste0(cvOutput, lf, '_q30_fixm_sortc_rd.bam')
  p1 = paste(in.s1, s3, s4, s5, s6, s7, sep=' ')
  writeLines(p1, oFile.param)
})

close(oFile.param)

oFile = file('AutoScripts/samtools.sh', 'wt')

writeLines(c(cvShell, cvJobName, cvArrayJob, cvNodes, cvProcessors, 
             cvRuntime, cvPartition, cvMemoryReserve, cvMail), oFile)

writeLines(c('# Autogenerated script from write_samtools_script.R', paste('# date', date())), oFile)
writeLines(c('# make sure directory paths exist before running script'), oFile)
writeLines('\n\n', oFile)

# module load
writeLines(c('module load samtools/1.13-gcc-9.4.0-python-3.8.12'), oFile)
writeLines('\n\n', oFile)

## write array job lines
writeLines("# Parse parameter file to get variables.
           number=$SLURM_ARRAY_TASK_ID
           paramfile=samtools_param.txt
           
           #ins1=`sed -n ${number}p $paramfile | awk '{print $1}'`
           bamfile=`sed -n ${number}p $paramfile | awk '{print $1}'`
           bamq30=`sed -n ${number}p $paramfile | awk '{print $2}'`
           bamq30sortn=`sed -n ${number}p $paramfile | awk '{print $3}'`
           bamq30sortnFixm=`sed -n ${number}p $paramfile | awk '{print $4}'`
           bamq30FixmSortc=`sed -n ${number}p $paramfile | awk '{print $5}'`
           bamrd=`sed -n ${number}p $paramfile | awk '{print $6}'`
           
           # 9. Run the program.", oFile)

# sam to bam
# p1 = paste('samtools view -b -S', '$ins1', '>', '$bamfile', sep=' ')
# com1 = paste(p1)
# remove low quality MAPQ reads
p1 = paste('samtools view -b -q 30', '$bamfile', '>', '$bamq30', sep=' ')
com2 = paste(p1)
# sort the file by name
p1 = paste('samtools sort -n -o', '$bamq30sortn', '$bamq30', sep=' ')
com3 = paste(p1)
# fixmate step
p1 = paste('samtools fixmate -m', '$bamq30sortn', '$bamq30sortnFixm', sep=' ')
com4 = paste(p1)
# sort file by coordinate
p1 = paste('samtools sort -o', '$bamq30FixmSortc', '$bamq30sortnFixm', sep=' ')
com5 = paste(p1)

# create index before duplicate removal
p1 = paste('samtools index', '$bamq30FixmSortc', sep=' ')
com5.1 = paste(p1)
# mark and remove duplicates, for paired end reads
p1 = paste('samtools markdup -r', '$bamq30FixmSortc', '$bamrd', sep=' ')
com6 = paste(p1)
# create index
p1 = paste('samtools index', '$bamrd', sep=' ')
com7 = paste(p1)

writeLines(c(com2, com3, com4, com5, com5.1, com6, com7), oFile)
writeLines('\n\n', oFile)

close(oFile)
